From: Myah Caron <qsniyg@protonmail.com>
Subject: [PATCH] ntdll: Avoid reading TEB in NtCreateThreadEx.
Message-Id: <jz1lSzzJ1CePLndEhs4FO9PWu-ZJ0Nc5XOQRtLTdvZNp6JzNFuYd-TtFUD8SgCqNJyhD4-5h5UENwMUREJqAZc14TwTLNN94DRa-kEJ3zBE=@protonmail.com>
Date: Mon, 28 Sep 2020 18:59:14 +0000

Wine-bug: https://bugs.winehq.org/show_bug.cgi?id=49903
Signed-off-by: Myah Caron <qsniyg@protonmail.com>
---
As noted in the bug, some programs hold a breakpoint on memory reads to the TEB, and they get triggered by GetCurrentProcessId. Since this is a syscall, this wouldn't occur under Windows.

 dlls/ntdll/unix/thread.c | 4 +++-
 server/protocol.def      | 1 +
 server/thread.c          | 1 +
 3 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/thread.c b/dlls/ntdll/unix/thread.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/thread.c
+++ b/dlls/ntdll/unix/thread.c
@@ -1252,6 +1252,7 @@ NTSTATUS WINAPI NtCreateThreadEx( HANDLE *handle, ACCESS_MASK access, OBJECT_ATT
     data_size_t len;
     struct object_attributes *objattr;
     struct ntdll_thread_data *thread_data;
+    DWORD pid = 0;
     DWORD tid = 0;
     int request_pipe[2];
     TEB *teb;
@@ -1315,6 +1316,7 @@ NTSTATUS WINAPI NtCreateThreadEx( HANDLE *handle, ACCESS_MASK access, OBJECT_ATT
         if (!(status = wine_server_call( req )))
         {
             *handle = wine_server_ptr_handle( reply->handle );
+            pid = reply->pid;
             tid = reply->tid;
         }
         close( request_pipe[0] );
@@ -1338,7 +1340,7 @@ NTSTATUS WINAPI NtCreateThreadEx( HANDLE *handle, ACCESS_MASK access, OBJECT_ATT
         goto done;
     }
 
-    set_thread_id( teb, GetCurrentProcessId(), tid );
+    set_thread_id( teb, pid, tid );
 
     thread_data = (struct ntdll_thread_data *)&teb->GdiTebBatch;
     thread_data->request_fd  = request_pipe[1];
diff --git a/server/protocol.def b/server/protocol.def
index 11111111111..11111111111 100644
--- a/server/protocol.def
+++ b/server/protocol.def
@@ -946,6 +946,7 @@ struct input_shared_memory
     int          request_fd;   /* fd for request pipe */
     VARARG(objattr,object_attributes); /* object attributes */
 @REPLY
+    process_id_t pid;          /* process id */
     thread_id_t  tid;          /* thread id */
     obj_handle_t handle;       /* thread handle (in the current process) */
 @END
diff --git a/server/thread.c b/server/thread.c
index 11111111111..11111111111 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -1814,6 +1814,7 @@ DECL_HANDLER(new_thread)
         thread->system_regs = current->system_regs;
         if (req->flags & THREAD_CREATE_FLAGS_CREATE_SUSPENDED) thread->suspend++;
         thread->dbg_hidden = !!(req->flags & THREAD_CREATE_FLAGS_HIDE_FROM_DEBUGGER);
+        reply->pid = get_process_id( process );
         reply->tid = get_thread_id( thread );
         if ((reply->handle = alloc_handle_no_access_check( current->process, thread,
                                                            req->access, objattr->attributes )))
